src/scope.ml

```diff
@@ -415,6 +420,19 @@ let scope_rule : sig_state -> p_rule -> sym * pp_hint * rule loc = fun ss r ->
     let t = scope (M_RHS(Array.to_list map)) ss Env.empty p_rhs in
     Bindlib.unbox (Bindlib.bind_mvar vars t)
   in
+  let _, t = Bindlib.unmbind rhs in
+  let rec privacy t =
+    (* Ensure that there are no foreign private symbols used. *)
+    let h, args = Basics.get_args t in
+    begin match unfold h with
+      | Symb({sym_path; sym_expo=Private; _},_) ->
+        if sym_path <> ss.signature.sign_path
+        then fatal p_lhs.pos "Foreign private symbol not allowed in RHS."
+      | _                                       -> ()
+    end;
+    List.iter privacy args
+  in
+  privacy t; (* Checking for private symbols in rhs. *)
```

Comment:
So, here, you make 2 traversals of the rhs. Moreover, you only check for private symbols in RHS. But my understanding is that private symbols should be checked every where except in LHS... Hence, my proposal to add some additional boolean argument to find_sym. In this case there is no need for a second traversal with this privacy function.

Reply:
Yes, it is not complete yet, I'll add an optional argument to find_sym which defaults to false